### PR TITLE
Fix #101: Rename `UserAuth` to `WebAuth`, mark `UserAuth` as deprecated, add `ApiAuth` authentication method

### LINKED ISSUE #101: Remove application layer logic from UserAuth::challenge()
Метод [UserAuth::challenge()](https://github.com/yiisoft/user/blob/896366d4c64fe65ee25d86d69957a9555cb92dc2/src/UserAuth.php#L39) должен возвращать переданный в него $response. Приложение должно само подготовить Response в соответствии с его логикой обработки неаутентифицированного пользователя в обрабтчике RequestHandlerInterface который конфигурируется в мидваре Authentication пакета yiisoft/auth через свойство [failureHandler](https://github.com/yiisoft/auth/blob/3dd4aef867509ed686f94de3ca9038e6ac978c74/src/Middleware/Authentication.php#L26).

Логика с 302 редиректом подходит для обычного приложения с рендерингом ответа на бэкэнде, но не подходит для SPA приложения.

Пример обработчика ответа
```
final class AuthRequestErrorHandler implements RequestHandlerInterface
{
    public function handle(ServerRequestInterface $request): ResponseInterface
    {
        return $this->responseFactory
            ->createResponse(Status::FOUND)
            ->withHeader('Location', $this->authUrl);
    }
}
```

di
```
Authentication::class => [
    'class' => Authentication::class,
    '__construct()' => [
        'authenticationFailureHandler' => Reference::to(AuthRequestErrorHandler::class),
    ],
],
```

### Additional info

| Q                | A
| ---------------- | ---
| Version          | 2.2.0

### PR DESCRIPTION
**Pull request by olegbaturin**
rename UserAuth to WebAuth
add ApiAuth

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌ has deprecated class

#101

### CODE REVIEW

File: config/di-web.php [hunk 1]
```diff
@@ -19,12 +19,11 @@
         },
     ],
 
-    UserAuth::class => [
-        'class' => UserAuth::class,
+    WebAuth::class => [
```
Comment (vjik) [inline config/di-web.php:22]: For BC need keep `UserAuth` definition also

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,7 +2,8 @@
 
 ## 2.2.1 under development
 
-- no changes in this release.
+- Chg #101: Rename `UserAuth` to `WebAuth`, mark `UserAuth` as deprecated (@olegbaturin)
+- Enh #101: Add `ApiAuth` authentication method (@olegbaturin)
```
Comment (vjik) [inline CHANGELOG.md:6]:
```suggestion
- New #101: Add `ApiAuth` authentication method (@olegbaturin)
```

**Review by vjik — CHANGES_REQUESTED** [review]

File: tests/WebAuthTest.php
Comment (vjik) [inline tests/WebAuthTest.php:1]: Keep `UserAuthTest` also.

**Review by vjik — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -190,12 +190,90 @@ The `Yiisoft\User\CurrentUser` instance is stateful, so when you build long-runn
...
+    AuthenticationMethodInterface::class => WebAuth::class,
+];
+```
+
+#### ApiAuth
+
+`Yiisoft\User\Method\ApiAuth` is used to authenticate users in the API clients.
+If authentication is failed, it returns the response from the `Yiisoft\Auth\Middleware\Authentication::authenticationFailureHandler` handler.
+
+In order to use `Yiisoft\User\Method\ApiAuth` as an authentication method, you need to define it as an implementation of `Yiisoft\Auth\AuthenticationMethodInterface` in the DI container configuration
```
Comment (samdark) [inline README.md:243]:
```suggestion
To use `Yiisoft\User\Method\ApiAuth` as an authentication method, you need to define it as an implementation of `Yiisoft\Auth\AuthenticationMethodInterface` in the DI container configuration
```

File: src/Method/ApiAuth.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+use Yiisoft\Auth\AuthenticationMethodInterface;
+use Yiisoft\Auth\IdentityInterface;
+use Yiisoft\User\CurrentUser;
+
+/**
+ * Implementation of the `AuthenticationMethodInterface` for authenticating users in the API clients.
+ */
+final class ApiAuth implements AuthenticationMethodInterface
+{
+    public function __construct(private CurrentUser $currentUser)
```
Comment (samdark) [inline src/Method/ApiAuth.php:18]:
```suggestion
    public function __construct(private readonly CurrentUser $currentUser)
```

File: src/Method/WebAuth.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+use Yiisoft\User\CurrentUser;
+
+/**
+ * Implementation of the `AuthenticationMethodInterface` for authenticating users in the web applications.
+ */
+final class WebAuth implements AuthenticationMethodInterface
+{
+    private string $authUrl = '/login';
+
+    public function __construct(private CurrentUser $currentUser, private ResponseFactoryInterface $responseFactory)
```
Comment (samdark) [inline src/Method/WebAuth.php:22]:
```suggestion
    public function __construct(
    	private readonly CurrentUser $currentUser,
    	private readonly ResponseFactoryInterface $responseFactory
	)
```

File: src/UserAuth.php [hunk 6]
```diff
@@ -13,6 +13,8 @@
 
 /**
  * Implementation of the authentication interface for the user.
+ *
+ * @deprecated Use {@see \Yiisoft\User\Method\WebAuth}. Will remove in the next major version.
```
Comment (samdark) [inline src/UserAuth.php:17]:
```suggestion
 * @deprecated Use {@see \Yiisoft\User\Method\WebAuth}. We will remove this class in the next major version.
```

File: tests/ApiAuthTest.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        $result = (new ApiAuth($user, new ResponseFactory()))->authenticate(new ServerRequest());
+
+        $this->assertNotNull($result);
+        $this->assertSame('test-id', $result->getId());
+    }
+
+    public function testIdentityNotAuthenticated(): void
+    {
+        $user = $this->createCurrentUser();
+        $result = (new ApiAuth($user, new ResponseFactory()))->authenticate(new ServerRequest());
```
Comment (samdark) [inline tests/ApiAuthTest.php:34]:
```suggestion
        $result = (new ApiAuth($user))->authenticate(new ServerRequest());
```

There's only one argument accepted, not two.

File: tests/ApiAuthTest.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+use Yiisoft\User\CurrentUser;
+use Yiisoft\User\Method\ApiAuth;
+
+final class ApiAuthTest extends TestCase
+{
+    public function testSuccessfulAuthentication(): void
+    {
+        $user = $this->createCurrentUser();
+        $user->login(new MockIdentity('test-id'));
+        $result = (new ApiAuth($user, new ResponseFactory()))->authenticate(new ServerRequest());
```
Comment (samdark) [inline tests/ApiAuthTest.php:25]:
```suggestion
        $result = (new ApiAuth($user))->authenticate(new ServerRequest());
```

File: tests/ApiAuthTest.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+        $this->assertNull($result);
+    }
+
+    public function testChallengeIsCorrect(): void
+    {
+        $response = new Response(Status::UNAUTHORIZED);
+        $user = $this->createCurrentUser();
+        $challenge = (new ApiAuth($user, new ResponseFactory()))->challenge($response);
```
Comment (samdark) [inline tests/ApiAuthTest.php:43]:
```suggestion
        $challenge = (new ApiAuth($user))->challenge($response);
```

**Review by samdark — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Pull request merged**